### ISSUE TITLE
'deis logs' not catching all app logs from docker containers

### ISSUE THREAD

**Issue by hivearts** [deis/deis#3633]
docker logs
```
Wed May 06 2015 07:44:45 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 07:45:13 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 07:47:21 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 07:47:50 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:06:08 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:06:37 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:08:58 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:09:27 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:25:35 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:26:02 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:29:41 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:30:09 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:41:58 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:42:26 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:48:18 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:48:46 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 09:01:14 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 09:01:43 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 09:05:35 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 09:06:03 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 09:08:18 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 09:08:47 GMT+0000 (UTC): Rebuilding website completed.
```

deis client logs
```
2015-05-06T09:54:15UTC vaer[web.1]: Wed May 06 2015 09:54:15 GMT+0000 (UTC): Building website to /public..
2015-05-06T09:54:46UTC vaer[web.1]: Wed May 06 2015 09:54:46 GMT+0000 (UTC): Building website completed.
2015-05-06T09:58:47UTC vaer[web.1]: Wed May 06 2015 09:58:47 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:07:39UTC vaer[web.1]: Wed May 06 2015 10:07:39 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:08:08UTC vaer[web.1]: Wed May 06 2015 10:08:08 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:08:56UTC vaer[web.1]: Wed May 06 2015 10:08:56 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:18:19UTC vaer[web.1]: Wed May 06 2015 10:18:19 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:19:08UTC vaer[web.1]: Wed May 06 2015 10:19:08 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:28:01UTC vaer[web.1]: Wed May 06 2015 10:28:01 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:28:32UTC vaer[web.1]: Wed May 06 2015 10:28:32 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:28:51UTC vaer[web.1]: Wed May 06 2015 10:28:51 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:29:21UTC vaer[web.1]: Wed May 06 2015 10:29:21 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:38:15UTC vaer[web.1]: Wed May 06 2015 10:38:15 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:38:43UTC vaer[web.1]: Wed May 06 2015 10:38:43 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:39:35UTC vaer[web.1]: Wed May 06 2015 10:39:35 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T11:53:31UTC deis[api]: hivearts deployed 99f07e0
2015-05-06T11:53:58UTC vaer[web.1]: Listening on port 5000
2015-05-06T11:53:58UTC vaer[web.1]: Wed May 06 2015 11:53:58 GMT+0000 (UTC): Building website to /public..
2015-05-06T11:54:28UTC vaer[web.1]: Wed May 06 2015 11:54:28 GMT+0000 (UTC): Building website completed.
2015-05-06T11:58:37UTC vaer[web.1]: Wed May 06 2015 11:58:37 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T12:00:46UTC vaer[web.1]: Wed May 06 2015 12:00:46 GMT+0000 (UTC): Rebuilding website completed.
```

As you can see deis not catching all logs from docker container, what should i do?
I am using v1.5.2 with v1.6.0-rc1 router.

**Comment by mboersma** [comment]
These two sets of log entries don't seem to overlap: the deis client logs are from a somewhat later time. Based on the docker logs, I assume we're expecting to see more `Webhook detected, rebuilding website started..` messages, but since these don't overlap it's hard to know for sure.

Do you have both logs from the same time series so we can verify that things are missing from `deis logs`?

**Comment by szymonpk** [comment]
I've noticed similar issue, you can see [log](https://gist.github.com/szymonpk/cdfa993acc2dbc5468a2) after scale web=8 & scale web=1. Nothing at all from web.8.

**Comment by hivearts** [comment]
Both are my last logs, i will try again later.
I will destroy app, deploy again and post here all logs.

**Comment by obmarg** [comment]
This has been happening to me regularly as well.  On 1.4.1 just now, but have witnessed it on older versions as well.  Sometimes it's just a couple of lines in the middle of a message missing, sometimes I never get logs at all from an instance.

Here's an example from right now.  When I run `deis logs` I get:

> 2015-05-07T01:13:53UTC connect-dev[web.1]: ERROR:root:Error in Taleo
2015-05-07T01:13:53UTC connect-dev[web.1]: Traceback (most recent call last):
2015-05-07T01:13:53UTC connect-dev[web.1]:   File "/app/connect/integrations/taleo_enterprise/ping.py", line 25, in ping
2015-05-07T01:13:53UTC connect-dev[web.1]:     client.get_usage()
2015-05-07T01:13:53UTC connect-dev[web.1]:   File "/app/raleo/enterprise/soap.py", line 288, in get_usage
2015-05-07T01:13:53UTC connect-dev[web.1]:     rv = self._call(self.ACTIONS['get_usage'], 'get_usage.xml')

But when I run journalctl I get:

> May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: ERROR:root:Error in Taleo
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: Traceback (most recent call last):
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: File "/app/connect/integrations/taleo_enterprise/ping.py", line 25, in ping
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: client.get_usage()
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: File "/app/raleo/enterprise/soap.py", line 288, in get_usage
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: rv = self._call(self.ACTIONS['get_usage'], 'get_usage.xml')
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: File "/app/raleo/enterprise/soap.py", line 197, in _call
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: raise TaleoSoapError(response.content)
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: raleo.enterprise.soap.TaleoSoapError: <?xml version='1.0' encoding='UTF-8'?><!DOCTYPE html PUBLIC "-//W3C//
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: &amp;wsdl">[wsdl]</a></li><li>RecruitmentSourceEventService ...
May 07 01:13:53 ip-10-21-2-91.us-west-1.compute.internal sh[5593]: 10.21.1.17 - - [07/May/2015:01:13:53 +0000] "POST /_admin/apsis/provision HTTP/1.1" 400 28 "-" "Python/3.4

Notice the 2 extra lines on the end there.

A similar request was made a few minutes later, and I got exactly the same lines omitted from the `deis logs` output - not sure if thats relevant, or just a coincidence.

I do notice this most often on python exception tracebacks - maybe something can't handle too much data being thrown at it at once...?

**Comment by bacongobbler** [comment]
that's likely the case here. Maybe the syslog server library we imported can't handle every incoming message? https://github.com/deis/deis/blob/master/logger/syslog/server.go#L99-L113

**Comment by bacongobbler** [comment]
@obmarg if you have it, logs from `deis-logger` might be helpful to see if the logs were discarded for any reason :)

**Comment by carmstrong** [comment]
@bacongobbler We should investigate this (and try to reproduce) for 1.7 - it seems several people have run into it.

**Comment by obmarg** [comment]
@bacongobbler 

Nothing much in the `journalctl -u deis-logger.service` output:

> -- Logs begin at Mon 2015-05-04 00:54:57 UTC, end at Thu 2015-05-07 11:04:14 UTC. --

But `docker logs deis-logger` gives me:

> Starting syslog...
Syslog server started...
deis-logger running
2015/04/29 08:41:50 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]
2015/04/29 08:41:50 501: All the given peers are not reachable (Tried to connect to each peer twice and failed) [0]

**Comment by bacongobbler** [comment]
Hmm, the only place where we use etcd in the logger would be for publishing and for retrieving the log drain. Both of those would not explain why the logger is failing to retrieve all logs from all containers.

When I scaled containers up to 5 and just curl'd 2000 times, all logs showed up instantly (using example-ruby-sinatra for this test):

```
><> for i in $(seq 1 2000); do curl --silent http://ruby.local3.deisapp.com >/dev/null; done
```

When I inspected /var/lib/deis/store/logs/ruby.log I saw that not all logs were there (note the `tail` is there to remove the first few logs that were not part of this exercise):

```
core@deis-01 /var/lib/deis/store/logs $ cat ruby.log | grep -e "-> /." | tail -n +49 | wc -l
1897
```

Still inspecting, but deis-logger doesn't seem to indicate anything, nor does deis-logspout.

**Comment by bacongobbler** [comment]
Note that we are using UDP which does see some packet loss, but research seems to show some users which only see packet losses of one in every 10,000 packets. There's probably missing logic that's magically hiding the fact that the packet was lost or something.

**Comment by mboersma** [comment]
Was this fixed with c73c8f7 or is there more to investigate here?

**Comment by bacongobbler** [comment]
I still think there's more to investigate here (https://github.com/deis/deis/issues/3633#issuecomment-106581975), though nothing from the logs seems to indicate an packet loss. I think I just need to dive into the logger/logspout and check if we're sending/receiving all logs from the socket, or if some go missing somewhere down the pipeline.

**Comment by LoicMahieu** [comment]
Confirmed in deis@1.7.3
App: https://github.com/LoicMahieu/deis-test-nodejs

deis logs:
```
$ deis logs
2015-08-02T17:42:22UTC deis[api]: loic created initial release
2015-08-02T17:42:48UTC deis[api]: loic deployed cbb1838
2015-08-02T17:42:48UTC deis[api]: loic scaled containers cmd=1
```

docker logs:
```
$ docker logs igloo-deis-nodejs-sample_v2.cmd.1

> igloo-deis-nodejs-sample@1.0.0 start /app
> node index.js

Listen to http://localhost:3000
10.10.0.235 - - [02/Aug/2015:17:43:28 +0000] "GET / HTTP/1.1" 200 11 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.107 Safari/537.36"
10.10.0.235 - - [02/Aug/2015:17:43:29 +0000] "GET / HTTP/1.1" 304 - "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.107 Safari/537.36"
```

I tried to destroy/re-create the app, scale up or down and always no logs.

**Comment by bacongobbler** [comment]
@LoicMahieu that looks more like logspout or logger is not logging properly, since the only logs are from the controller. Would you mind opening a separate issue? Thanks@

**Comment by LoicMahieu** [comment]
@bacongobbler Moved to sperate issue

**Comment by krancour** [comment]
I believe this may be a problem of scale.  I had previously observed in #4264 that I wasn't getting app logs when using k8s as the scheduler.  Context-- that was observed after having deployed, _significantly scaled_, and destroyed multiple apps... rinsed, repeated.

I deployed a _new_ app.  No logs.  After restarting logspout, I started getting logs again.  Then I scaled that app up to 200 containers... parts of my logs went missing right then.

Observe-- logs before scaling:

```
2015-08-13T21:22:12UTC deis[api]: kent created initial release
2015-08-13T21:24:26UTC deis[api]: kent deployed 14fb753
2015-08-13T21:24:27UTC deis[api]: kent scaled containers web=1
2015-08-13T21:25:15UTC quaint-vacation[web]: 2015/08/13 21:25:15 listening on 5000...
2015-08-13T21:26:49UTC quaint-vacation[web]: 2015/08/13 21:26:49 10.244.7.0:58924 GET /
...
2015-08-13T21:27:19UTC quaint-vacation[web]: 2015/08/13 21:27:19 10.244.7.0:58994 GET /
```

Logs after scaling:

```
2015-08-13T21:37:39UTC deis[api]: kent scaled containers web=200
2015-08-13T21:37:41UTC quaint-vacation[web]: 2015/08/13 21:37:41 listening on 5000...
...
2015-08-13T21:38:38UTC quaint-vacation[web]: 2015/08/13 21:38:38 listening on 5000...
```

Skip ahead to the punch line:

```
[kent@mbp example-go]$ deis logs | grep -c "listening on 5000"
71
```

71 != 200

Even though I can verify all 200 are running:

```
[kent@mbp example-go]$ kubectl get pods --all-namespaces | grep -c Running
200
```

At this stage, it appears very much that logging is left only partially working.  I make 1000 requests:

```
[kent@mbp example-go]$ deis logs -n 10000 | grep -c "GET /"
520
[kent@mbp example-go]$ for i in `seq 1 1000`; do echo $i; curl http://quaint-vacation.krancour.deis.ninja/ ; done
# output omitted for brevity
[kent@mbp example-go]$ deis logs -n 10000 | grep -c "GET /"
987
```

987 - 520 != 1000

Destroy the app, re-create, without scaling past 1, and hit it a few times:

```
[kent@mbp example-go]$ deis logs
2015-08-13T22:09:32UTC deis[api]: kent created initial release
2015-08-13T22:11:38UTC deis[api]: kent deployed 14fb753
2015-08-13T22:11:39UTC deis[api]: kent scaled containers web=1
```

^ Logs only from the controller and none from the app!

If I scale up to 200 again, I get some messages from the app, but fewer then before.  After scaling back down to 1 and back up to 200 again-- nothing.  I completely beat logspout into submission.

```
2015-08-13T22:18:48UTC deis[api]: kent scaled containers web=200
2015-08-13T22:18:50UTC gifted-undertow[web]: 2015/08/13 22:18:50 listening on 5000...
2015-08-13T22:18:51UTC gifted-undertow[web]: 2015/08/13 22:18:51 listening on 5000...
2015-08-13T22:18:51UTC gifted-undertow[web]: 2015/08/13 22:18:51 listening on 5000...
2015-08-13T22:18:53UTC gifted-undertow[web]: 2015/08/13 22:18:52 listening on 5000...
2015-08-13T22:18:54UTC gifted-undertow[web]: 2015/08/13 22:18:54 listening on 5000...
2015-08-13T22:18:54UTC gifted-undertow[web]: 2015/08/13 22:18:54 listening on 5000...
2015-08-13T22:18:55UTC gifted-undertow[web]: 2015/08/13 22:18:55 listening on 5000...
2015-08-13T22:19:04UTC gifted-undertow[web]: 2015/08/13 22:19:04 listening on 5000...
2015-08-13T22:21:15UTC deis[api]: kent scaled containers web=1
2015-08-13T22:22:18UTC deis[api]: kent scaled containers web=200
2015-08-13T22:26:41UTC deis[api]: kent scaled containers web=1
```

^ No more "listening on 5000" _at all_.

I don't want to imply this is the _only_ problem, but it is definitely _a_ problem.

**Comment by bacongobbler** [comment]
@krancour do you have logs of the deis-logspout agents as well as deis-logger? That sounds exactly like what I saw in https://github.com/deis/deis/issues/3633#issuecomment-106581975.

**Comment by krancour** [comment]
@bacongobbler I just repeated the whole experiment and kept a close eye on logspout's own logs throughout.  There are absolutely no error messages.  What is telling though is what's _not_ there.  When logspout is functioning, you get messages that look like this as container's are added:

```
Aug 13 23:04:13 ip-10-21-2-229.ec2.internal sh[31506]: 2015/08/13 23:04:13 attach: a1dda243d2fc k8s_POD.e4cc795_gifted-undertow-v2-web-5gi4h_gifted-undertow_9533ef95-420f-11e5-8910-12b0e808152f_df8a2ca1 success
```

And messages like this when such a container goes away:

```
Aug 13 23:06:54 ip-10-21-2-229.ec2.internal sh[31506]: 2015/08/13 23:06:54 attach: a1dda243d2fc finished
```

As logspout begins to tap out, you don't see those anymore.

**Comment by bacongobbler** [comment]
good find. Thanks @krancour, I'll see what I can dig up :)

**Comment by krancour** [comment]
I'm reviewing logspout code to try and pinpoint the cause as well... tomorrow.

Worth noting... at some point we forked and customized logspout code within the deis repo.  Logspout itself has evolved quite a bit since then.  I don't know if it's worth attempting to upgrade to see if that improves the situation at all.

**Comment by bacongobbler** [comment]
The only modification we made to logspout was to make it [self aware of etcd](https://github.com/deis/deis/blob/master/logspout/logspout.go#L175-L189) for service discovery of the logger as well as a [custom datetime format](https://github.com/deis/deis/blob/master/logspout/logspout.go#L73). It shouldn't be too hard to re-implement if we wanted to try bumping logspout (famous last words)

**Comment by krancour** [comment]
I'm willing to give that a go.

**Comment by krancour** [comment]
Note that my experiment with scaling up and down repeatedly is only one way to create and destroy many containers.  An accumulation of releases of an app running at smaller scale could also knock out logspout over time.

**Comment by krancour** [comment]
I ran a variety of experiments to try and recreate this issue today under different (simpler) circumstances in order to isolate the problem.  I won't enumerate all the different things I tried, but they are many.  In _all but one_ experiment, I was unable to replicate the underlying issue of logspout eventually appearing to "give up" on attaching to new containers.  The thing all those experiments had in common was that they weren't actually shipping the logs anywhere.  When I finally started shipping the logs somewhere (like `deis-logspout` ordinarily does), the problem came right back.

You'd think that attaching to the container and shipping the logs are two separate concerns, but maybe not.  Here's my hypothesis: if the UDP connection to the log server fails, the process of attaching to a container is failing silently.

netstat (inside the logspout container) is showing me a pretty variable number of active udp connections, but never more than the number of containers that logspout is attached to (or should be attached to).  So it doesn't seem to me that there's a real problem of connections being left open or anything.  More likely, in my mind, there's an issue with the golang code that's brokering these connections.

That's what I'll be on the hunt for now, but any insight from anyone and/or an extra set of eyes is welcome.

**Comment by krancour** [comment]
@bacongobbler fwiw, I just tried substituting the latest gliderlabs/logspout for our forked version and it didn't curb this behavior at all.  So whatever is causing this, it has _not_ been fixed between when we forked it and now.

**Comment by krancour** [comment]
Also, I've discovered that I can't seem to recreate this problem just by launching and killing many containers from a script that uses the Docker CLI.  I've been able to recreate the issue with Fleet as the scheduler, but not reliably.  I can make it happen quite reliably with k8s as the scheduler.  I don't think the problem is isolated to k8s, but I think the frequency and ease with which I can replicate the issue using k8s may be a meaningful clue.
